doc/designs/app-passwords.md

```diff
@@ -0,0 +1,312 @@
+SSSD.
+
+## Implementation
+
+The extension must be backward compatible, that means not breaking any existing
+functionality and passing all existing tests.
+
+As for changes to the data storage:
+* As described above, a new LDAP subtree will be created for app passwords, with
+  objects holding them being grouped by app, and their attributes being, besides
+  an app password itself for LDAP, a uid, a descriptive display name, an app
+  name, a UUID, and a backlink to the owning user.
+* New ACIs will be created so that users will have needed access to objects
+  holding their own app passwords.
+
+A new objectclass will be created for the `apppw` object and added into
+the LDAP schema. This objectclass will be defined in
+`install/share/60basev3.ldif`:
+```
+objectClasses:
+    (2.16.840.1.113730.3.8.12.XX
+    NAME 'ipaAppPw'
+    DESC 'Object containing an app password and related data'
+    SUP top
+    MUST ( uid $ displayName $ appName $ entryUUID $ managedBy )
+    X-ORIGIN 'IPA v4.9')
+```
+
+The `userPassword` attribute will be added by including the
+`simpleSecurityObject` into the `apppw` object.
+
+The LDIF code to add the `cn=apps,cn=accounts,$SUFFIX` entry, as well as new
+ACIs for access to the `apppw` object and its attributes, during
+* new installations will be in `install/share/apppw.ldif`,
+* upgrades will be in `install/updates/19-apppw.update` (entry) and
+  `install/updates/20-aci.update` (ACIs).
+
+There will be the following ACIs on the `cn=apps,cn=accounts,$SUFFIX` entry:
+```
+aci: (target = "cn=\*,cn=apps,cn=accounts,$SUFFIX")
+     (version 3.0; acl "Allow users to add entries grouping app passwords for a
+     particular app";
+     allow (add) 
+     userdn = "ldap:///all;)
+aci: (target = "ldap:///uid=($dn)-\*,cn=\*,cn=apps,cn=accounts,$SUFFIX")
+     (targetattr = "uid || displayName || userPassword || appName || entryUUID
+     || managedBy")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to add an app password for themselves";
+     allow (add) 
+     userdn = "ldap:///uid=($dn),cn=users,cn=accounts,$SUFFIX";)
+aci: (targetattr = "managedBy")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to search for their app passwords";
+     allow (search) 
+     userdn = "ldap:///self" OR
+     userattr = "managedBy#USERDN";)
+aci: (targetattr = "displayName")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to read display names for their app
+     passwords";
+     allow (read) 
+     userdn = "ldap:///self" OR
+     userattr = "managedBy#USERDN";)
+aci: (target = "ldap:///uid=\*,cn=\*,cn=apps,cn=accounts,$SUFFIX")
+     (targetattr = "uid || displayName || userPassword || appName || entryUUID
+     || managedBy")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to delete their app password";
+     allow (delete) 
+     userdn = "ldap:///self" OR
+     userattr = "managedBy#USERDN";)
+```
```

Comment:
You probably want to add ipaAppPw to all the other password-based ACIs

Reply:
Never mind, it's covered by userPassword already.

Reply:
What do you mean? ipaAppPw is a proposed new structural objectclass specifically for objects representing app passwords. It is not meant for other passwords. I also don't know why would we want to expand some other existing ACIs to apply also to app passwords - if this is the case, can you explain reasons for this, please?

Reply:
As I said, I misread it, it's fine as-is.

Reply:
OK, I apologize - again, I did not see your second post. I got distracted and then did not refresh the page before submitting my response.